qiime2/core/archive/provenance.py

```diff
@@ -317,7 +317,12 @@ def finalize(self, final_path, node_members):
         self.write_action_yaml()
         self.write_citations_bib()
 
-        self.path.rename(final_path)
+        # Certain networked filesystems will experience a race
+        # condition on `rename`, so fall back to copying.
+        try:
+            self.path.rename(final_path)
+        except FileExistsError:
+            distutils.dir_util.copy_tree(str(self.path), str(final_path))
```

Comment:
I think `self.path` needs to be updated to be the `final_path` and the original needs to be destroyed.

Reply:
Well, it's possible `self.path` isn't used after this, but I think we're still going to have a dangling directory in /tmp/

Reply:
Yeah I thought we already had dangling dirs hanging around in some other cases, so I wasn't too worried about cleaning them up, but I can do that here if you want. I was kinda thinking that would be punted for v3 of this code, though.

Reply:
I suppose I just remember the horror of working out the lifetimes for all of these objects and so I don't want to lose whatever logic was implied by the `rename`

Reply:
under normal conditions, the framework won't leave behind any dangling files. Now I've definitely seen this fail with certain errors, but never looked to closely

Reply:
> Now I've definitely seen this fail with certain errors, but never looked to closely

Lol, yeah, me too - I think the race condition implied in this PR is one of those cases!

I'll add some code to clean up! Also, good catch on the `self.path` update - I completely spaced that.

Thanks for the review!

Reply:
I ran into some issues with updating `self.path`, but ran out of time this morning to look into it further. If you get a second to peek at it, please do.

Reply:
Ok, looking at the semantics of `rename`, it never actually changed state, so `self.path` can remain stale (it is clearly never used again). Sorry about that, the docs weren't super clear so I tested it on Py3.6 and a `rename` only seems to side-effect the system, but not the object.